Refactoring of automated cucumber tests

al-niessner opened the pull request:
## 🗒️ Summary
Complete rewrite of cucumber testing.

## ⚙️ Test Data and/or Report
All unit tests should pass and there should be 251 cucumber tests.

## ♻️ Related Issues
Closes #1102

Commit:
giant rewrite of cucumber

Reduced the content in any given line to just the essentials. Use of quotes in cucumber is odd but workable. Allowing some columns to be blank was not intuitive but is also workable. The line is much shorter with little to know repetitive information. The checks are more complete and StepDefs is completely cleaned up. Weill, probably not completely but good enoough for next step.

nutjob4life reviewed (approved):
Visual inspection: ✓
Maven tests: ✓
```
[INFO] Results:
[INFO] 
[WARNING] Tests run: 25, Failures: 0, Errors: 0, Skipped: 23
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

jordanpadams commented on src/test/resources/features/3.6.x.feature:
```diff
@@ -1,76 +1,9 @@
-    Then produced output from validate command should be similar to reference <refOutputValue> or no error reported.
-
-    @v3.6.x
+Feature: 3.6.x
```
@al-niessner looks great! a couple comments:

* will need a `testId` column in the feature file. this ties to our I&T system and ensure we have a unique identifier for each test that traces back to ticket. @tloubrieu-jpl can maybe provide some more details about what is expected there, but I am pretty sure a `testId` column is needed.
* I would prefer we keep the tags in the files, if possible, so you can run `mvn test` for a specific subset of tests, e.g. `-Dcucumber.filter.tags='@pre.v3.6.x'`
* @tloubrieu-jpl, do we need the a specific feature name associated with these tests for TestRail, or does it not matter?

al-niessner replied:
@jordanpadams 

1. testId still exists but comes in two parts: column 1 and column 2 if first index is 1. The [scenerio outline](https://github.com/NASA-PDS/validate/blob/63712705c64cc25066d7ddc5ef21c0f635539328/src/test/resources/features/3.6.x.feature#L2) contains all of the boilerplate and joins those two columns. In eclipse it looks like:  
```
Scenario Outline: NASA-PDS/validate#1028 1                                                                                                                                     [90m# src/test/resources/features/3.6.x.feature:9[0m
  [32mGiven [0m[32man [0m[32m[1m1028[0m[32m, [0m[32m[1m1[0m[32m, and [0m[32m[1m"github1028"[0m                                                                                                                                           [90m# cucumber.StepDefs.an_and(java.lang.Integer,java.lang.Integer,java.lang.String)[0m
  [32mWhen [0m[32mexecute validate with [0m[32m[1m"--skip-context-validation -t {datasrc}/xa.s16..shz.1976.070.0.xml --schema {datasrc}/PDS4_PDS_1N00.xsd --schematron {datasrc}/PDS4_PDS_1N00.sch"[0m [90m# cucumber.StepDefs.execute_validate(java.lang.String)[0m
  [32mThen [0m[32mcompare to the .[0m                                                                                                                                                        [90m# cucumber.StepDefs.compare_to_the()[0m

Scenario Outline: NASA-PDS/validate#1028 2                                                                                                                                     [90m# src/test/resources/features/pre.3.6.x.feature:9[0m
  [32mGiven [0m[32man [0m[32m[1m1028[0m[32m, [0m[32m[1m2[0m[32m, and [0m[32m[1m"github1028"[0m                                                                                                                                           [90m# cucumber.StepDefs.an_and(java.lang.Integer,java.lang.Integer,java.lang.String)[0m
  [32mWhen [0m[32mexecute validate with [0m[32m[1m"--skip-context-validation -t {datasrc}/xa.s16..shz.1976.070.0.xml --schema {datasrc}/PDS4_PDS_1N00.xsd --schematron {datasrc}/PDS4_PDS_1N00.sch"[0m [90m# cucumber.StepDefs.execute_validate(java.lang.String)[0m
  [32mThen [0m[32mcompare to the .[0m                                                                                                                                                        [90m# cucumber.StepDefs.compare_to_the()[0m
```

If you look at the output from [github actions](https://github.com/NASA-PDS/validate/actions/runs/12798025411/job/35681201596?pr=1113) you will see the same thing:
```
Scenario Outline: NASA-PDS/validate#1028 1                                                                                                                                     # src/test/resources/features/3.6.x.feature:9
  Given an 1028, 1, and "github1028"                                                                                                                                           # cucumber.StepDefs.an_and(java.lang.Integer,java.lang.Integer,java.lang.String)
  When execute validate with "--skip-context-validation -t {datasrc}/xa.s16..shz.1976.070.0.xml --schema {datasrc}/PDS4_PDS_1N00.xsd --schematron {datasrc}/PDS4_PDS_1N00.sch" # cucumber.StepDefs.execute_validate(java.lang.String)
  Then compare to the .                                                                                                                                                        # cucumber.StepDefs.compare_to_the()

Scenario Outline: NASA-PDS/validate#1028 1                                                                                                                                     # src/test/resources/features/pre.3.6.x.feature:9
  Given an 1028, 1, and "github1028"                                                                                                                                           # cucumber.StepDefs.an_and(java.lang.Integer,java.lang.Integer,java.lang.String)
  When execute validate with "--skip-context-validation -t {datasrc}/xa.s16..shz.1976.070.0.xml --schema {datasrc}/PDS4_PDS_1N00.xsd --schematron {datasrc}/PDS4_PDS_1N00.sch" # cucumber.StepDefs.execute_validate(java.lang.String)
  Then compare to the .                                                                                                                                                        # cucumber.StepDefs.compare_to_the()
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5.888 s -- in cucumber.CucumberTest
```

Yeah, I updated my pre.3.6.x to be subtest 2 while both are 1 in the commited files. This problem exists no matter how you do it and can use a python script to check the files that the same ticket number and subtest do not exist.

al-niessner replied:
@jordanpadams 

2. I did not know there was a purpose to that redundant information. Add it back.

Commit:
minor updates based on comments

Commit:
first translation of feature files

Commit:
little cleanup translation of feature files

Commit:
a little more cleanup

Commit:
merge from main

Commit:
no going back now

Commit:
another fix

Commit:
another cleanup step

Commit:
try using given catalog

Commit:
another fix

Commit:
maybe last set of fixes

Commit:
all tests should pass

Bulk of the work is done. Made the report.json match (give or take) for the tests from when run on main as well. If they are wrong, they are as wrong as they ere anyway.

Added a new test function for running single scenerios as though typed in on the command line. cut-n-paste is no longer your friend when translating scenerios to command lines. However, what it should look like is printed out so you can just grab it from there now.

Commit:
removed local file and cleaned up expectation

Commit:
catch up to modern day validate

Commit:
update some helper java

jordanpadams committed:
Add underscore to scenario outline

jordanpadams committed:
Update the scenario setup

jordanpadams committed:
Fix bug in feature file

Commit:
cleaned up expectatiosn

Three of the tests were trying to download or access network resources that were not available during the recreation of the feature files. These tests had some erroneous expectations because of the external lack of resources. While the fault is clear it does demonstrate a weakness of the tests that they can fail for reasons beyond the control of the validate repo.
